### PR TITLE
Only mark unions as uninhabited if all of their fields are uninhabited

### LINKED ISSUE #46845: Unions with uninhabited variants are considered uninhabited
Uninhabited variants should ideally be irrelevant to the representation, just like in enums.

As a test, this prints out 0 instead of 8:
```rust
#![feature(never_type)]
use std::mem;

union Foo {
    a: u64,
    b: !
}

fn main() {
    println!("{}", mem::size_of::<Foo>());
}
```
This happens because rustc considers unions more like structs than enums, and the [test for uninhabitedness](https://github.com/rust-lang/rust/blob/b39c4bc12358078f77ddd01288b24252f757f37d/src/librustc/ty/layout.rs#L1361-L1364) sees a single variant with two fields, one of which is uninhabited.

I have a fix for this, but it's slightly easier to bundle it with some other changes I'm working on.

### PR DESCRIPTION
**Pull request by gereeter**
Fixes #46845.

### CODE REVIEW

**Comment by eddyb** [comment]
r? @eddyb

cc @arielb1 Should we do this or should we just never consider unions uninhabited?

**Comment by scottmcm** [comment]
What should `ManuallyDrop<!>` do?  This reminds me of the "does it need a `()` field" discussion from back in https://github.com/rust-lang/rust/pull/40559#issuecomment-287279145...

**Comment by gereeter** [comment]
I would expect `ManuallyDrop<!>` to be uninhabited both just intuitively and because if I'm using `ManuallyDrop`, I'm probably doing it for optimization purposes, so taking an optimization away seems problematic. If unions with no inhabited fields are inhabited, then there are certain (admittedly fairly minor, probably) optimization that just can't be done. I don't think a `()` field in `ManuallyDrop` is necessary because `drop_in_place` already leaves its target invalid and I would expect that to be true of the method on `ManuallyDrop` as well.

That's not to say that there isn't a use for a union
```rust
union MaybeInitialized<T> {
    init: T,
    uninit: ()
}
```
That should be used whenever cleaned up data sticks around, that is when "uninitialized" is a valid state to be in.

All that said, I think `arrayvec` is a good example of why this might not be desirable. Currently, an `ArrayVec<[T; 20]>` is represented as `(ManuallyDrop<[T; 20]>, usize)`. This is totally invalid if `ManuallyDrop<!>` is uninhabited. I personally believe it is invalid anyway, since the type does not at all encode the fact that the `[T; 20]` might be *partially* initialized. Even if `ManuallyDrop` had a `()` variant, I would expect any union sanitizer to object to this code - the union is in the initialized state, but the inner value is invalid. However, at the very least, it shows that people today are writing code that assumes that values stored in unions can be invalid.

The safe way to write `ArrayVec`, under the stricter reasoning I'm arguing for, would be to use an associated type parameter so that `<[T; 20]>::PartiallyInitialized == [MaybeInitialized<T>; 20]` and store the `PartiallyInitialized` version of an array.

**Comment by eddyb** [comment]
@gereeter Part of the problem is that people have collapsed `MaybeInitialized` and `ManuallyDrop`.

**Comment by gereeter** [comment]
@eddyb Oh, absolutely. I guess my point is that people have gone further and collapsed some `PartiallyInitialized` and `ManuallyDrop`, so I'm not sure that declaring `MaybeInitialized` to be the same thing as `ManuallyDrop` will actually help. Therefore, I'd rather just declare that code incorrect.

**Comment by eddyb** [comment]
@gereeter I still prefer not having unions able to be uninhabited but maybe there's a point to it?

**Comment by scottmcm** [comment]
Hmm, what's an empty union? In the "a union is an enum without an explicit discriminant" model it's uninhabited; in the "a union is a struct where all the fields overlap" model it's a ZST.

Though in both of those models, a union with one uninhabited field is uninhabited...

**Comment by gereeter** [comment]
I admit that the biggest motivation for me is that it feels more consistent. We call enums tagged unions, but an enum with all uninhabited variants is uninhabited. Therefore, a tag with a union with all uninhabited variants should be uninhabited. We have safe functions like `ManuallyDrop::into_inner`. Therefore, if `ManuallyDrop<T>` is inhabited, we can get a `T`, so `T` is inhabited. Also, as I said above, I shouldn't lose an optimization just because of an implementation detail, and uninhabitedness is a useful optimization.

**Comment by gereeter** [comment]
@scottmcm Is the following a valid implementation of `transmute`?
```rust
union Transmuter<T, U> {
    start: T,
    end: U
}

unsafe fn transmute<T, U>(value: T) -> U {
    Transmuter { start: value }.end
}
```

In the "overlapping struct" model, I'd say definitely yes. In the "tagless enum" model, I'm not sure.

**Comment by eddyb** [comment]
> just because of an implementation detail

This isn't really about implementation details, which don't really matter here either way.
The real question is the model that we actually want to expose to the user.

**Comment by gereeter** [comment]
@eddyb Sorry, I meant an implementation detail of a data structure library. If I have a data structure like `NonEmptyList<T>` and I want it to be uninhabited if `T` is uninhabited, then I am *forbidden* from using `T` only inside of a union (if unions are always inhabited). Unless `PhantomData` induces uninhabitedness, but that seems very wrong to me. Therefore, the optimization I want to give to the users of my library forces my hand for implementation.

**Comment by eddyb** [comment]
@gereeter What does a sound usecase look like? If you can optimize it out, how do you use `union`?

**Comment by scottmcm** [comment]
Uninhabitedness is the first thing I've seen where the "a union is an enum" model is clearly better than the"a union is a struct" model, since a struct is uninhabited if any field is uninhabited (aka the issue at hand) whereas an enum is uninhabited only if all variants are uninhabited.

@gereeter That's a question for the smantics-of-unsafe-code folks, I think -- it seems open as of https://github.com/petrochenkov/rfcs/blob/8f1a960844b6ae37ec19fd89d39355fa98b1bb2b/text/1444-union.md#delayed-and-unresolved-questions

**Comment by arielb1** [comment]
I think that we want unions (or, at least, `ManuallyDrop`) to always be inhabited - unions are used in unsafe code, and introducing basically unneeded and randomly-firing footguns into unsafe code is a stupid thing.

**Comment by gereeter** [comment]
@eddyb I'm thinking of some sort of wrapper like
```rust
struct PermutedDrop<A: Array> {
    data: [ManuallyDrop<A::Item>; A::LENGTH],
    order: [usize; A::LENGTH]
}

impl<A: Array> Drop for PermutedDrop<A> {
    fn drop(&mut self) {
        for idx in self.order {
            unsafe {
                ManuallyDrop::drop(&mut self.data[idx]);
            }
        }
    }
}
```
that rearranges the drop order of the things it contains. If `T` is uninhabited, `PermutedDrop<[T; 20]>` should also be uninhabited. If unions are always uninhabited, I believe that `PermutedDrop` cannot be written to work and have that property. I admit, however, that this is a very niche use case where the optimization is unlikely to matter - why would you put uninhabited things in `PermutedDrop` anyway?

@arielb1 I'd posit that almost all code that assumes that unions cannot be uninhabited is broken in some other way. In that case, this being a footgun is not a problem - it just makes it more likely that the unsoundness will be noticed. That said, I may be wrong about this, and I think the situation is a bit different with `ManuallyDrop`, so it seems reasonable to at least make `ManuallyDrop` always be inhabited.

**Comment by eddyb** [comment]
I don't even think we treat `[!; N]` as uninhabited, just ZST.

**Comment by gereeter** [comment]
@eddyb Oh. Well, that also seems wrong to me (for `N` greater than 0). However, it does largely invalidate my argument.

Since it seems to be the most accepted opinion, doesn't break code, and isn't setting a precedent that hasn't already been set, I switched this PR to always making unions inhabited.

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -1352,22 +1352,14 @@ impl<'a, 'tcx> LayoutDetails {
...
-                }
-
                 if def.is_union() {
                     let packed = def.repr.packed();
                     if packed && def.repr.align > 0 {
                         bug!("Union cannot be packed and aligned");
                     }
+                    if variants.len() != 1 {
+                        bug!("Union must be represented as a single variant");
+                    }
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: Can you remove this too?

**Comment by kennytm** [comment]
@bors retry 

```
[01:49:45] test net::tcp::tests::clone_accept_smoke has been running for over 60 seconds

No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

(`check x86_64-apple-darwin`)

**Comment by alexcrichton** [comment]
Looks like we forgot to backport this to 1.23.0 (sorry about that!) so removing the beta tags

**Comment by eddyb** [comment]
@alexcrichton #45225 was backed out from that beta (IIUC) so there wasn't anything to backport.

**Comment by alexcrichton** [comment]
Oh yay!

**Pull request merged**